### DISCUSSION TITLE
SearchFilter gives error in Full-text search using ('@') even when using Django and PostgreSQL

### DISCUSSION THREAD

**Discussion by kushshiv** [encode/django-rest-framework#8205] [Question & Answer] [answered]
## Issue :
SearchFilter gives error in Full-text search using ('@').

## Documentation : 
[Documentation link](https://www.django-rest-framework.org/api-guide/filtering/#searchfilter)

>The search behavior may be restricted by prepending various characters to the search_fields.
>
>'^' Starts-with search.
>'=' Exact matches.
>'@' Full-text search. (Currently only supported Django's PostgreSQL backend.)
>'$' Regex search.

## Error:
FieldError(\"Unsupported lookup 'search' for CharField or join on the field not permitted.\")

**Comment by lovelydinosaur** [comment]
There's not enough information here for anyone to be able to help you.

**Reply by kushshiv** [reply]
I have an API to create, list and retrieve a Moment. Also have added a search, filter and order capability in the API. 
I am using (Django 3.2.4) and (PostgreSQL - 13.1).
According to the documentation I can do full text search in the API using `@` : 
>'@' Full-text search. (Currently only supported Django's PostgreSQL backend.)
I am getting below error when I do the search. 

Below is the code  where I have used `@name` in `search_fields`: 
```
class MomentViewSet(viewsets.GenericViewSet):
    """
    Moment Viewset is used to create, update and retrieve a moment, and list all the moments.
    """
    queryset = Moment.objects.all()
    serializer_class = MomentSerializer
    filter_backends = [DjangoFilterBackend, SearchFilter, OrderingFilter]
    filterset_fields = ['name']
    search_fields = ['@name']
    ordering_fields = ['name']


    def list(self, request: Request):
        """
        Get the list of all the moments.
        """
        try:
            moment_objects = self.filter_queryset(self.get_queryset())
            moment_serializer = MomentSerializer(moment_objects, many=True)
            moment_data = moment_serializer.data
            return Response(moment_data, status=status.HTTP_200_OK)

        except Exception as err:
            response = ReturnValue(False,
                                   f"Error while getting list of moments",
                                   error=repr(err)).to_dict()
            return Response(response, status=status.HTTP_503_SERVICE_UNAVAILABLE)
```
If I hit the endpoint :
`http://127.0.0.1:8000/moment/?search=office`
I am getting error :
```
HTTP 503 Service Unavailable
Allow: GET, POST, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept

{
    "success": false,
    "message": "Error while getting list of moments",
    "error": "FieldError(\"Unsupported lookup 'search' for CharField or join on the field not permitted.\")",
    "data": []
}
```
As per the documentation it should work in Django and postgres. But still getting this error.

**Reply by lovelydinosaur** [reply]
I think you want `name=office` there, maybe?

**Reply by kushshiv** [reply]
That will be for the filtering part... According to documentation here for SearchFilter, I can do search the on the fields specified in the `search_fields`. Please correct if I am doing something wrong.
[Search Filter Document](https://www.django-rest-framework.org/api-guide/filtering/#searchfilter)
As per document : 
>This will allow the client to filter the items in the list by making queries such as:
>
>http://example.com/api/users?search=russell

**Reply by lovelydinosaur** [reply]
Oh yes. Perhaps the two different filters you're using here are conflicting? I can't understand the intent behind having both `DjangoFilterBackend` *and* `SearchFilter` here.

Might be you'd want to swap the ordering of them around...

```python
filter_backends = [SearchFilter, DjangoFilterBackend, OrderingFilter]
```

**Reply by kushshiv** [reply]
Ok. This time I tried with only SearchFilter, and search_fields... Still the same issue...
 
```
queryset = Moment.objects.all()
serializer_class = MomentSerializer
filter_backends = [SearchFilter]
search_fields = ['@name']
```

**Reply by lovelydinosaur** [reply]
>  Still the same issue...

I'd double check that - the "Unsupported lookup" error is from django-filter...

https://github.com/carltongibson/django-filter/blob/fe90e3a5fdeaff0983d1325a3e9dcf3458ef078f/django_filters/exceptions.py

**Reply by kushshiv** [reply]
Finally got it. It will only work if `django.contrib.postgres` is added in the `INSTALLED_APPS`.

https://docs.djangoproject.com/en/3.2/ref/contrib/postgres/search/#the-search-lookup

>To use the search lookup, 'django.contrib.postgres' must be in your INSTALLED_APPS.

It would be highly appreciated if mention in the same [document](https://www.django-rest-framework.org/api-guide/filtering/#searchfilter) as well. 
PS : There was one more issue raised for the same.  https://github.com/encode/django-rest-framework/issues/5795

**Comment by kushshiv** [comment, marked as answer]
It will only work if django.contrib.postgres is added in the INSTALLED_APPS.

https://docs.djangoproject.com/en/3.2/ref/contrib/postgres/search/#the-search-lookup

>To use the search lookup, 'django.contrib.postgres' must be in your INSTALLED_APPS.

**Comment by sevdog** [comment]
On django admin site is it possible to use a `SearchVectorField` for searches by putting `field__exact` in `search_fields`, from [django docs](https://docs.djangoproject.com/en/3.2/ref/contrib/admin/#django.contrib.admin.ModelAdmin.search_fields): 

> If you don’t want to use **`icontains`** as the lookup, you can use any lookup by appending it the field. For example, you could use **`exact`** by setting **`search_fields`** to **`['first_name__exact']`**.

I expected the `SearchFilter` to behave like that, but as of now is not. If you try to use any lookup it does not work at all.

It would be a good thing if `SearchFilter.construct_search` was inspired by the admin method:

https://github.com/django/django/blob/107865780aa44914e21d27fdf4ca269bc61c7f01/django/contrib/admin/options.py#L1132-L1159

**Reply by auvipy** [reply]
drf is mainly based on django ORM. and contrib apps, but I don't think it follow much of django Admin patterns. I could be wrong though

**Reply by sevdog** [reply]
If you inspect the sources both of `SearchFilter` and `django.contrib.admin` search implementation they are too much lookalike to assert that `SearchFilter` is not "derived" from it (apart that in django.contrib some undocumented methods like `models.Q.create`).
Also looking at [`SearchFilter` docs](https://www.django-rest-framework.org/api-guide/filtering/#searchfilter) it exposes quite the same programming API of [`ModelAdmin.search_fields`](https://docs.djangoproject.com/en/3.2/ref/contrib/admin/#django.contrib.admin.ModelAdmin.search_fields) (especially the _prefix-lookup_ mechanism, ie: `^field` which means `field__istartswith` in both).

Last but not the least, it is stated clearly:
> The `SearchFilter` class supports simple single query parameter based searching, and is based on the [Django admin's search functionality](https://docs.djangoproject.com/en/stable/ref/contrib/admin/#django.contrib.admin.ModelAdmin.search_fields).

Thus it should also follow the lookups like admin site does.

**Reply by auvipy** [reply]
I checked your PR and I'm convinced that it could be a nice compat improvement :dancers:
